python/ray/autoscaler/_private/gcp/node_provider.py

```diff
@@ -68,206 +67,132 @@ def method_with_retries(self, *args, **kwargs):
 
 
 class GCPNodeProvider(NodeProvider):
-    def __init__(self, provider_config, cluster_name):
+    def __init__(self, provider_config: dict, cluster_name: str):
         NodeProvider.__init__(self, provider_config, cluster_name)
-
         self.lock = RLock()
         self._construct_clients()
 
         # Cache of node objects from the last nodes() call. This avoids
         # excessive DescribeInstances requests.
-        self.cached_nodes = {}
+        self.cached_nodes: Dict[str, GCPNode] = {}
 
     def _construct_clients(self):
-        _, _, self.compute = construct_clients_from_provider_config(
+        _, _, compute, tpu = construct_clients_from_provider_config(
             self.provider_config)
 
+        # Dict of different resources provided by GCP.
+        # At this moment - Compute and TPUs
+        self.resources: Dict[GCPNodeType, GCPResource] = {}
+
+        self.resources[GCPNodeType.COMPUTE] = GCPCompute(
+            compute, self.provider_config["project_id"],
+            self.provider_config["availability_zone"], self.cluster_name)
+        self.resources[GCPNodeType.TPU] = GCPTPU(
+            tpu, self.provider_config["project_id"],
+            self.provider_config["availability_zone"], self.cluster_name)
+
```

Comment:
I would also avoid constructing the TPU client if it is not necessary. The Compute client is probably always necessary because the head node is (presumably) a compute instance.

Reply:
Yeah, that's a good point, I realized that later too. I'll make it so it's not constructed unless necessary